pkg/datapath/linux/node.go

```diff
@@ -1446,6 +1470,101 @@ func (n *linuxNodeHandler) NodeNeighborRefresh(ctx context.Context, nodeToRefres
 	}
 }
 
+// NodeCleanNeighbors cleans all neighbor entries for previously used network
+// link interfaces. It should be used when the agent changes the state from
+// `n.enableNeighDiscovery = true` to `n.enableNeighDiscovery = false`.
+func (n *linuxNodeHandler) NodeCleanNeighbors() {
+	devToClean, err := loadNeighLinks(option.Config.StateDir)
+	if err != nil {
+		log.WithError(err).Error("Unable to load old neigh links configuration." +
+			" Previous running state needs to be cleaned up manually.")
+		return
+	}
+	if len(devToClean) == 0 {
+		return
+	}
+	// Delete the file after cleaning up neighbor list.
+	defer os.Remove(filepath.Join(option.Config.StateDir, neighListFileName))
+
+	n.mutex.Lock()
```

Comment:
Why is this lock needed?

Reply:
answered here https://github.com/cilium/cilium/pull/15993/files#r627610921

Reply:
I'm not following - we get the entries from the kernel not from the cilium-agent state, and there is only one routine which will be removing those entries. From what we are protecting?